python/ray/data/datasource/iceberg_datasource.py

```diff
@@ -0,0 +1,201 @@
+
+        # From largest to smallest, add the plan files ot the smallest chunk one at a
+        # time
+        for plan_file in sorted(plan_files, key=lambda f: f.length, reverse=True):
+            smallest_chunk = min(chunk_sizes, key=chunk_sizes.get)
+            chunks[smallest_chunk].append(plan_file)
+            # Update the size of the chunk after adding the file to it
+            chunk_sizes[smallest_chunk] += plan_file.length
+
+        return chunks
+
+    def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
+        from pyiceberg.io import pyarrow as pyi_pa_io
+
+        def _get_read_task(
+            tasks: Iterable["FileScanTask"],
+            table_identifier: str,
+            schema: "Schema",
+        ) -> Iterable[Block]:
+            # Closure so we can pass this callable as an argument to ReadTask
+
+            # Both the catalog and tbl attributes cannot be pickled, which means they
+            # must be instantiated within this function (as opposed to being attributes
+            # of the IcebergDatasource class)
+            catalog = self._get_catalog()
+            tbl = catalog.load_table(table_identifier)
+
+            # Use the PyIceberg API to read only a single task (specifically, a
+            # FileScanTask) - note that this is not as simple as reading a single
+            # parquet file, as there might be delete files, etc. associated, so we
+            # must use the PyIceberg API for the projection.
+            yield pyi_pa_io.project_table(
+                tasks=tasks,
+                table=tbl,
+                row_filter=self._row_filter,
+                projected_schema=schema,
+                case_sensitive=self._scan_kwargs.get("case_sensitive", True),
+                limit=self._scan_kwargs.get("limit"),
+            )
+
+        # Get the PyIceberg scan
+        data_scan = self._get_data_scan()
+        # Get the plan files in this query
+        plan_files = self.plan_files
+
+        # Get the projected schema for this scan, given all the row filters,
+        # snapshot ID, etc.
+        projected_schema = data_scan.projection()
+        # Get the arrow schema, to set in the metadata
+        pya_schema = pyi_pa_io.schema_to_pyarrow(projected_schema)
+
+        # Set the n_chunks to the min of the number of plan files and the actual
+        # requested n_chunks, so that there are no empty tasks
+        if parallelism > len(list(plan_files)):
+            parallelism = len(list(plan_files))
+            logger.warning(
+                f"Reducing the parallelism to {parallelism}, as that is the"
+                "number of files"
+            )
+
+        read_tasks = []
+        # Chunk the plan files based on the requested parallelism
+        for chunk_tasks in IcebergDatasource._distribute_tasks_into_equal_chunks(
+            plan_files, parallelism
+        ):
+            metadata = BlockMetadata(
+                num_rows=None,
```

Comment:
is it feasible to get the number of rows and add it to BlockMetadata here? it will allow for much better progress reporting, as we are moving to have Ray Data's default progress bar reporting be in `rows/s`.